core/src/cluster_info_vote_listener.rs

```diff
@@ -515,40 +527,65 @@ impl ClusterInfoVoteListener {
                             continue;
                         }
 
-                        // Don't insert if we already have marked down this pubkey
-                        // voting for this slot
-                        let maybe_slot_tracker =
-                            all_slot_trackers.read().unwrap().get(&slot).cloned();
-                        if let Some(slot_tracker) = maybe_slot_tracker {
-                            if slot_tracker.read().unwrap().voted.contains(vote_pubkey) {
-                                continue;
-                            }
-                        }
                         let unduplicated_pubkey = vote_tracker.keys.get_or_insert(vote_pubkey);
-                        diff.entry(slot).or_default().insert(unduplicated_pubkey);
+                        diff.entry(slot)
+                            .or_default()
+                            .insert(unduplicated_pubkey, true);
                     }
 
                     subscriptions.notify_vote(&vote);
-                    let _ = verified_vote_sender.send((*vote_pubkey, vote));
+                    let _ = verified_vote_sender.send((*vote_pubkey, vote.slots));
                 }
             }
         }
 
-        for (slot, slot_diff) in diff {
+        // Process the replay votes
+        for votes in replay_votes {
+            for (pubkey, slot) in votes.iter() {
+                if *slot <= root {
+                    continue;
+                }
+                let unduplicated_pubkey = vote_tracker.keys.get_or_insert(pubkey);
+                diff.entry(*slot)
+                    .or_default()
+                    .entry(unduplicated_pubkey)
+                    .or_default();
+            }
+        }
+
+        for (slot, mut slot_diff) in diff {
             let slot_tracker = vote_tracker
                 .slot_vote_trackers
                 .read()
                 .unwrap()
                 .get(&slot)
                 .cloned();
             if let Some(slot_tracker) = slot_tracker {
+                {
+                    let r_slot_tracker = slot_tracker.read().unwrap();
+                    // Only keep the pubkeys we haven't seen voting for this slot
+                    slot_diff.retain(|pubkey, is_from_gossip| {
+                        let is_new = !r_slot_tracker.voted.contains(pubkey);
+                        if is_new && !*is_from_gossip {
+                            let _ = verified_vote_sender.send((**pubkey, vec![slot]));
```

Comment:
This logic is essentially notification for votes received through replay similar to existing logic:
https://github.com/solana-labs/solana/blob/master/core/src/cluster_info_vote_listener.rs#L531-L532

Because replay votes will have a lot of duplicates (it's sending the same slots that it sees from `VoteState`'s in from every bank in ReplayStage) we deduplicate them here.

@CriesofCarrots note I didn't call `subscriptions.notify_vote(&vote);` because we don't have the actual vote transaction + hash from ReplayStage, just a slot in a `VoteState`

Reply:
@CriesofCarrots, comment added here: https://github.com/solana-labs/solana/pull/11119/commits/9b424ef7cdbee831d57d0cf4f6ed42883eb30a53